src/Microsoft.TemplateEngine.Cli/TemplateResolution/TemplateResolver.cs

```diff
@@ -485,5 +443,197 @@ private static void AddParameterMatchingToTemplates(IReadOnlyCollection<ITemplat
+                    }
+
+                    hasShortNamePartialMatch |= shortNameIndex > -1;
+                }
+
+                if (nameIndex > -1)
+                {
+                    return new MatchInfo { Location = MatchLocation.Name, Kind = MatchKind.Partial };
+                }
+
+                if (hasShortNamePartialMatch)
+                {
+                    return new MatchInfo { Location = MatchLocation.ShortName, Kind = MatchKind.Partial };
+                }
+
+                return new MatchInfo { Location = MatchLocation.Name, Kind = MatchKind.Mismatch };
+            };
+        }
+
+        /// <summary>
+        /// In addition to <see cref="ITemplateInfo"/> the class contains <see cref="TemplateInfoWithGroupShortNames.GroupShortNameList"/> property which contains the short names of other templates in the template group.
+        /// The class is used for template filtering using specific <see cref="CliNameFilter(string)"/> filter which takes into account the short names of template group when matching names.
+        /// </summary>
+        private class TemplateInfoWithGroupShortNames : ITemplateInfo, IShortNameList
+        {
+            private TemplateInfoWithGroupShortNames (ITemplateInfo source)
+            {
+                Author = source.Author;
+                Description = source.Description;
+                Classifications = source.Classifications;
+                DefaultName = source.DefaultName;
+                Identity = source.Identity;
+                GeneratorId = source.GeneratorId;
+                GroupIdentity = source.GroupIdentity;
+                Precedence = source.Precedence;
+                Name = source.Name;
+                ShortName = source.ShortName;
+                Tags = source.Tags;
+                CacheParameters = source.CacheParameters;
+                Parameters = source.Parameters;
+                MountPointUri = source.MountPointUri;
+                ConfigPlace = source.ConfigPlace;
+                LocaleConfigPlace = source.LocaleConfigPlace;
+                HostConfigPlace = source.HostConfigPlace;
+                ThirdPartyNotices = source.ThirdPartyNotices;
+                BaselineInfo = source.BaselineInfo;
+                HasScriptRunningPostActions = source.HasScriptRunningPostActions;
+            }
+
+            internal static TemplateInfoWithGroupShortNames FromITemplateInfo(ITemplateInfo source)
+            {
+                TemplateInfoWithGroupShortNames filterableTemplate = new TemplateInfoWithGroupShortNames(source);
+
+                if (source is IShortNameList sourceWithShortNameList)
+                {
+                    filterableTemplate.ShortNameList = sourceWithShortNameList.ShortNameList;
+                }
+                return filterableTemplate;
+            }
+
+            public string Author { get; private set; }
```

Comment:
All `private set;` in this class could be removed, to ensure immutability of properties also inside this class.